packages/auth/src/AuthProvider/AuthProvider.tsx

```diff
@@ -0,0 +1,186 @@
+
+//   return null
+// }
+
+type AuthProviderProps = {
+  type: string
+  skipFetchCurrentUser?: boolean
+  children: ReactNode
+}
+
+export function createAuthProvider<
+  TUser,
+  TRestoreAuth,
+  TLogIn,
+  TLogOut,
+  TSignUp,
+  TForgotPassword,
+  TResetPassword,
+  TValidateResetToken,
+  TVerifyOtp
+>(
+  AuthContext: React.Context<
+    | AuthContextInterface<
+        TUser,
+        TLogIn,
+        TLogOut,
+        TSignUp,
+        TForgotPassword,
+        TResetPassword,
+        TValidateResetToken
+      >
+    | undefined
+  >,
+  authImplementation: AuthImplementation<
+    TUser,
+    TRestoreAuth,
+    TLogIn,
+    TLogOut,
+    TSignUp,
+    TForgotPassword,
+    TResetPassword,
+    TValidateResetToken,
+    TVerifyOtp
+  >
+) {
+  /**
+   * @example
+   * ```js
+   *  const client = new Auth0Client(options)
+   *  // ...
+   *  <AuthProvider client={client} type="auth0" skipFetchCurrentUser={true}>
+   *    {children}
+   *  </AuthProvider>
+   * ```
+   */
+  const AuthProvider = ({
+    type,
+    children,
+    skipFetchCurrentUser,
+  }: AuthProviderProps) => {
+    // const [hasRestoredState, setHasRestoredState] = useState(false)
+
+    const [authProviderState, setAuthProviderState] = useState<
+      AuthProviderState<TUser>
+    >(defaultAuthProviderState)
+
+    // const [authImplementation, setAuthImplementation] =
+    //   useState<AuthImplementation>()
+
+    // const rwClientPromise: Promise<AuthClient> = useMemo(async () => {
+    //   // If ever we rebuild the rwClient, we need to re-restore the state.
+    //   // This is not desired behavior, but may happen if for some reason the host app's
+    //   // auth configuration changes mid-flight.
+    //   setHasRestoredState(false)
+
+    //   const rwClient = await createAuthClient(client, type, config)
+
+    //   setRwClient(rwClient)
+
+    //   return rwClient
+    // }, [client, type, config])
+
+    /**
+     * Clients should always return null or token string.
+     * It is expected that they catch any errors internally.
+     * This catch is a last resort effort in case any errors are
+     * missed or slip through.
+     */
+    const getToken = useToken(authImplementation)
+    const getCurrentUser = useCurrentUser(authImplementation)
+    const reauthenticate = useReauthenticate(
+      authImplementation,
+      setAuthProviderState,
+      skipFetchCurrentUser
+    )
+    const hasRole = useHasRole(authProviderState.currentUser)
+    const signUp = useSignUp(
+      authImplementation,
+      setAuthProviderState,
+      skipFetchCurrentUser
+    )
+    const logIn = useLogIn(
+      authImplementation,
+      setAuthProviderState,
+      skipFetchCurrentUser
+    )
+    const logOut = useLogOut(authImplementation, setAuthProviderState)
+    const forgotPassword = useForgotPassword(authImplementation)
+    const resetPassword = useResetPassword(authImplementation)
+    const validateResetToken = useValidateResetToken(authImplementation)
```

Comment:
@Tobbe circling back from yesterday I think to combine my work and your work we would just have to take all of this and shove it in a `useClient` hook. So instead of doing it directly in the AuthProvider, the user has to do it on their end. That would give the power of multiple auth providers to the user. And we would just have to change the props of `AuthProvider`. Doing it this way I think would also allow us to automatically generate `useAuth` without having to export a `useAuth` and `AuthProvider` from each package. Not sure if that messes with types or not. But maybe we can hop on a chat to discuss more.